spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java

```diff
@@ -719,11 +717,24 @@ private void setPartition(MessageBuilder<?> builder, Message<?> source) {
 		}
 	}
 
+	private void setKey(MessageBuilder<?> builder, Message<?> source) {
+		Object key = getReplyKeyFromRequest(source);
+		// Set the reply record key only for non-batch requests
+		if (key != null && !(key instanceof List)) {
+			builder.setHeader(KafkaHeaders.KEY, key);
+		}
+	}
+
 	@Nullable
 	private byte[] getReplyPartition(Message<?> source) {
 		return source.getHeaders().get(KafkaHeaders.REPLY_PARTITION, byte[].class);
 	}
 
+	@Nullable
+	private Object getReplyKeyFromRequest(Message<?> source) {
+		return source.getHeaders().get(KafkaHeaders.RECEIVED_KEY);
```

Comment:
I don't see a reason in this one-line method.
More over it feel like `static`.
Way, either way: I'd prefer to not have it making me thinking "where else this method can be used" 😄